managed-pointer/cpp-20/D0567R1.md

```diff
@@ -5,222 +5,384 @@
+When attempting to standardise a unified interface for data movement there are many differeent aspects which need to be considered. These come from both the various architectures that this interface aims to abstract and the various data flow models which this interface aims to provide.
 
-### Naming considerations
+1. Must support different memory architectures
+2. Must support different levels of synchronisation
+3. Must support different data flow models
+4. Must integrate seemlessly with exexcutors
+5. Must be interoperable between executor implementations
+6. Must allow optimisation of data movement between devices
+7. Must abstract device capabilities
 
-During the development of this paper, many names were considered both for the `managed_ptr` itself and for its interface.
+This paper in it's current state aims to satisfy [1], [3], [4], [5], [6], [7], while [2] is still being developed.
 
-Alternative names that were considered for `managed_ptr` were `temporal` as it described a container which gave temporal access to a managed memory allocation, `managed_container` as the original design was based on the `std::vector` container and `managed_array` as the managed memory allocation is statically sized.
+### 2.1. Must support different memory architectures
 
-Alternative names for the `put()` and `get()` interface were `acquire()` and `release()` as you were effectively acquiring and releasing the managed memory allocation and `send()` and `receive()` as you are effectively sending and receiving back the managed memory allocation.
+In non-heterogeneous or non-distributed systems, there is typically a single device: a host CPU with a single memory region or an area of addressable memory. In contrast, heterogeneous and distributed systems have multiple devices (including the host CPU) often with their own discrete memory regions. A device in this respect can be any architecture that exists within a system that is C\+\+ programmable; this can include CPUs, GPGPUs, APUs, FPGAs, DSPs, NUMA nodes, I/O devices and other forms of accelerators.
+
+These memory regions can be categorised into two groups. Firstly memory architectures which have a single unfied address space, that is a single area of addressable memory that is accessible across multiple devices, this can optionally be supprted in hardware by devices having the same physical memory or in software by having a cache coherancy layer accessing a shared virtual address space. Secondly memory architectures which have multiple address spaces, therefore having separate discrete areas of addresable memory, often in different physical memory.
+
+A unified interface needs to be capable of supporting both of these kinds of memory architectures in such a way that any requirements for cache coherancy is abstracted by the interface and associated coherency model.
```

Comment:
So this is something that is being considered for systems with shared memory regions, however this is not yet covered by the proposal. This is currently described in the future work section in "5.1. Shared Accessibility".